osquery/dispatcher/scheduler.cpp

```diff
@@ -184,15 +187,25 @@ void SchedulerRunner::start() {
     if ((i % 3) == 0) {
       relayStatusLogs(true);
     }
-
-    // Put the thread into an interruptible sleep without a config instance.
-    pauseMilli(interval_ * 1000);
+    auto loop_step_duration =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::steady_clock::now() - start_time_point);
+    if (loop_step_duration < interval_millisec) {
+      // Put the thread into an interruptible sleep without a config instance.
+      pauseMilli(interval_millisec - loop_step_duration);
+    } else {
+      sum_overtime_ += loop_step_duration - interval_millisec;
+    }
```

Comment:
How about 
if ((loop_step_duration + sum_overtime_ )  < interval_millisec) {
  pauseMilli(interval_millisec - loop_step_duration - sum_overtime_ );
} else {      
  sum_overtime_ += loop_step_duration - interval_millisec;
}

Reply:
Yes, it can make the thread to work without resting. What do we call this, super fast mode?

Reply:
I don't think it is a desirable behavior. Consider the case, when your query takes a long time or during the loop step the lid of laptop was closed. It means `sum_overtime_` will grow up and loop will no sleep for a pretty long time. Which means for a while we will have a CPU greedy process and many the same queries will be runned in one moment. But it is wrong. If we need to see the process list every 2 seconds, the results taken for every 50miliseconds is wrong, right?

Reply:
Yes you are right. I forgot about system going to freeze.

Reply:
You both have good points, can we have some kind of compromise here? Something like the scheduler tries to recover from drift smaller than 1 minute? I'm also not super happy that if we have expensive queries we'll still drift event if those queries only run infrequently. @akindyakov I see your point but I would say that if a user is scheduling queries every 2 seconds and at the same time has expensive queries, than that's something we currently not support. This may require you to keep two overtime counters, and do the loop slightly different than what @uptycs-nishant suggested (like when 0 < overtime < 60 instead of sleep do overtime -= loop_step_duration and continue) but would remove drift for most use cases.

Reply:
I am correcting myself uptime seems to give correct results on linux and macos vms, but correct source of time which takes care of "system going for freeze" on linux and macos is -

struct timespec ts;
clock_gettime(CLOCK_MONOTONIC,&ts);
On windows I still need to check. This is posix call maybe it is supported on windows as well.

Reply:
Maybe getting uptime is a good way to go. Is something like `std::chrono` able to abstract that for us? We just need to make sure that getting the uptime is not somehow more expensive than getting time because we're going to use it a lot.

Also I still think you still need to use the `overtime -= (interval_millisec - loop_step_duration)` logic (my initial comment had this wrong), because queries might take more than 1 second in which case `(loop_step_duration + sum_overtime_ ) < interval_millisec)` will always be false and you'll never recover, right?

Reply:
std::chrono::steady_clock is wrapper over clock_gettime(CLOCK_MONOTONIC,&ts) (on linux I verified )
From the cpp reference "Class std::chrono::steady_clock represents a monotonic clock".
So I dont think we need to change anything here and worry about "system going to freeze in the middle of the query execution" No need to to use uptime since that does not take care of freeze or suspended time.  

> Also I still think you still need to use the overtime -= (interval_millisec - loop_step_duration) logic (my initial comment had this wrong), 
We are doing already doing that in the else part which is -
sum_overtime_ += loop_step_duration - interval_millisec; right?

> because queries might take more than 1 second in which case (loop_step_duration + sum_overtime_ ) < interval_millisec) will always be false and you'll never recover, right?

There will be times when no query will be running in an iteration. That time the condition (loop_step_duration + sum_overtime_ ) < interval_millisec) will help scheduler to recover from drift. It depends now how much we want to recover.

Reply:
Ok, I agree. Let recover from `summary_drift_` using sleep time. If it is ok sometimes to run some particular query more often than it is mentioned in schedule.

Are you agree to organize sleep condition like this?
```c++
if (sum_overtime_ > std::chrono::seconds{60} || loop_step_duration + sum_overtime_ < interval_millisec) {
  sleep(interval_millisec - loop_step_duration - sum_overtime_);
  sum_overtime_ = 0;
} else {
  sum_overtime_ += loop_step_duration - interval_millisec;
}
```
This solution gives up if summary drift too big to manage it.
If loop duration plus summary drift is less than interval, so we can afford to sleep. e.g. summary drift is negative for some reason or `loop_step_duration` is small enough.
Otherwise it skips the sleep().

Reply:
if (sum_overtime_ > std::chrono::seconds{60} || loop_step_duration + sum_overtime_ < interval_millisec) {
//  value (interval_millisec - loop_step_duration - sum_overtime_) can be negative. So following statement can go for log sleep.
  sleep(interval_millisec - loop_step_duration - sum_overtime_); 
  sum_overtime_ = 0;
} else {
  sum_overtime_ += loop_step_duration - interval_millisec;
}

Re wrote the above logic. 
if (loop_step_duration + sum_overtime_ <= interval_millisec) {
   // We are on track.
   sleep(interval_millisec - loop_step_duration - sum_overtime_);
   sum_overtime_ = 0;
} else if (sum_overtime_ > std::chrono::seconds{60}) {
  // Giving up. Taking some rest.
  sleep(interval_millisec/2);
  sum_overtime_ = 0;
} else {
  // Recover from drift.
  sum_overtime_ += loop_step_duration - interval_millisec;
}

Reply:
Yes, you are right - we should not sleep for a negative time :) Thank you!

But, your line:
```c++
sleep(interval_millisec/2);
```
looks like a magic, it is absolutely not clear why it is only half of interval. I think to 'give up' is enough to put a zero to `sum_overtime_`. Like this:
```c++
if (sum_overtime_ > std::chrono::seconds{60}) {
  // giving up
  sum_overtime_ = 0;
}
if (loop_step_duration + sum_overtime_ <= interval_millisec) {
  sleep(interval_millisec - loop_step_duration - sum_overtime_);
  sum_overtime_ = 0;
} else {
  sum_overtime_ += loop_step_duration - interval_millisec;
}
```
Fair enough?

Reply:
My thought was after doing overtime thread needs to yield the cpu for some time(so that it should not show spike). Earlier I thought it should be at least 1 second then settled for sleep(interval_millisec/2);

Reply:
Assuming that not every step of the loop is overloaded we can be sure that after setting up `sub_overtime_ = 0` loop will take some rest (on this step or on the next one), right?

Reply:
Yes we can drop this.